how-to/workspace-platform-starter/client/src/modules/actions/favorites-menu/actions.ts

```diff
@@ -59,56 +67,95 @@ export class FavoritesMenuProvider implements Actions {
+					if (!isEmpty(client)) {
+						const favInfo = await client.getInfo();
+						const menuEntries: PopupMenuEntry<FavoriteEntry>[] = [];
 
-							const menuEntries =
-								favorites?.map((f) => ({
-									label: f.label ?? "",
-									icon: f.icon,
-									customData: f
-								})) ?? [];
+						const iconFolder = await this._helpers.getCurrentIconFolder();
+						const colorScheme = await this._helpers.getCurrentColorSchemeMode();
 
-							const result = await this._helpers.showPopupMenu<FavoriteEntry>(
-								{ x: payload.x, y: 48 },
-								payload.windowIdentity,
-								`There are no favorite ${type}s`,
-								menuEntries,
-								{
-									mode: "native"
+						if (favInfo.enabledTypes) {
+							let hadEntries = false;
+							for (const type of favInfo.enabledTypes) {
+								const saved = await client.getSavedFavorites(type);
+								if (saved && saved.length > 0) {
+									if (hadEntries) {
+										menuEntries.push({ type: "separator" });
+									}
+									menuEntries.push(
+										...saved
+											.sort((f1, f2) => (f1.label ?? "").localeCompare(f2.label ?? ""))
+											.map((f) => ({
+												label: f.label ?? "",
+												icon: this.themeUrl(f.icon, iconFolder, colorScheme),
+												customData: f
+											}))
+									);
+									hadEntries = true;
 								}
-							);
+							}
+						}
 
-							if (isEmpty(result)) {
-								this._logger?.info("Favorites Menu Dismissed");
-							} else {
-								this._logger?.info("Favorites Menu Item Selected", result);
+						const result = await this._helpers.showPopupMenu<FavoriteEntry>(
+							{ x: payload.x - 16, y: 40 },
+							payload.windowIdentity,
+							"There are no favorites",
+							menuEntries,
+							{
+								mode: this._settings?.menuType ?? "native"
+							}
+						);
 
-								if (result.type === FAVORITE_TYPE_NAME_APP) {
-									if (!isEmpty(this._helpers?.launchApp)) {
-										await this._helpers?.launchApp(result.typeId);
-									}
-								} else {
-									this._logger?.info(`Favorites Type ${result.type} no yet supported`, result);
+						if (isEmpty(result)) {
+							this._logger?.info("Favorites Menu Dismissed");
+						} else {
+							this._logger?.info("Favorites Menu Item Selected", result);
+
+							if (result.type === FAVORITE_TYPE_NAME_APP) {
+								if (!isEmpty(this._helpers?.launchApp)) {
+									await this._helpers?.launchApp(result.typeId);
+								}
+							} else if (result.type === FAVORITE_TYPE_NAME_PAGE) {
+								if (!isEmpty(this._helpers?.launchPage)) {
+									const page = await platform.Storage.getPage(result.typeId);
+									await this._helpers?.launchPage(page);
 								}
+							} else if (result.type === FAVORITE_TYPE_NAME_WORKSPACE) {
+								if (!isEmpty(this._helpers?.launchWorkspace)) {
+									const workspace = await platform.Storage.getWorkspace(result.typeId);
+									await this._helpers?.launchWorkspace(workspace);
```

Comment:
I made a comment up above about making the abstraction more useful to the platform dev by limiting options for the module while still making the module developer's life easier by removing the need to use the platform.Storage apis). This also means the platform doesn't have to trust the workspace object passed as it is doing the lookup itself directly.

Reply:
Done